### PR TITLE
More Cypress end-to-end tests

### PR DESCRIPTION
**Pull request by emilvirkki**
Adds a comprehensive end-to-end test suite for Reki.

Merging #271 first will make this easier to review, as otherwise this PR is pretty big.

### CODE REVIEW

File: src/client/components/ParticipantListPage/containers/ParticipantCount.tsx [hunk 1]
```diff
@@ -5,7 +5,7 @@ const Count: React.FC<{ count: number | undefined }> = ({ count }) => {
   if (count === undefined) {
     return <Spinner />
   } else {
-    return <>count</>
+    return <>{count}</>
```
Comment (googol) [inline src/client/components/ParticipantListPage/containers/ParticipantCount.tsx:8]: oops :sweat_smile:

File: src/server/boot/03-partioid-login.js [hunk 2]
```diff
@@ -17,7 +17,9 @@ export default function (app) {
   const partioIdEntryPoint = `https://${partioIDRemoteName}.partio.fi/simplesaml/saml2/idp/SSOService.php`
   const partioIdLogoutUrl = `https://${partioIDRemoteName}.partio.fi/simplesaml/saml2/idp/SingleLogoutService.php`
   const partioIdCertificate = fs.readFileSync(
-    path.resolve(`./certs/partioid/${partioIDRemoteName}.crt`),
+    path.resolve(
+      `${__dirname}/../../../certs/partioid/${partioIDRemoteName}.crt`,
```
Comment (googol) [inline src/server/boot/03-partioid-login.js:21]: is this required to make the tests work? In general I prefer not using `__dirname`, since it makes the path fragile to refactorings (if the call site is moved to a different file or the file is moved to a different directory)

File: test/e2e/specs/authentication-spec.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+describe('Authentication', () => {
+  beforeEach(() => {
+    cy.task('createUser', {
+      email: 'masa',
```
Comment (googol) [inline test/e2e/specs/authentication-spec.js:4]: should these email and username be extracted into variables within the describe context?

File: test/e2e/specs/participants-spec.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      password: 'Salasana123',
+    })
+    cy.task('loadFixtures', 'participants-and-deps.json')
+    cy.login('masa', 'Salasana123')
+    cy.contains(/Leiriläiset/i).click()
+  })
+
+  // REKI shouldn't show results before searching to prevent users from accidentally
+  // seeing personally identifiable information
+  it('are not shown in the list initially', () => {
```
Comment (googol) [inline test/e2e/specs/participants-spec.js:14]: this is good, should we also have one that tests clearing the search, to make sure that the list resets to zero

File: test/utils/test-utils.js [hunk 5]
```diff
@@ -78,6 +84,15 @@ export async function deleteFixturesIfExist(modelName, whereClause) {
   return models[modelName].destroy({ where: whereClause || {} })
 }
 
+export async function deleteAllFixtures() {
+  for (const model of Object.keys(models)) {
```
Comment (googol) [inline test/utils/test-utils.js:88]: this whole thing could read clearer as:
```ts
await Promise.all(Object.keys(models).filter(m => m !== 'UserRole').map(deleteFixturesIfExist))
```

ymmv

File: test/utils/test-utils.js [hunk 6]
```diff
@@ -70,6 +70,12 @@ export function createFixture(modelName, fixture) {
   }
 }
 
+export async function createFixtures(fixtures) {
+  for (const model in fixtures) {
```
Comment (googol) [inline test/utils/test-utils.js:74]: this could be just:
```ts
await Promise.all(Object.entries(fixtures).map(([model, fixture]) => createFixture(model, fixture)))
```

**Review by googol — COMMENTED** [review]
this is looking pretty good, a couple of comments

File: src/server/boot/03-partioid-login.js (hunk 2 above)
Comment (emilvirkki) [reply src/server/boot/03-partioid-login.js:21]: Unfortunately yes - it's somehow related to the Cypress TypeScript config having a different current working directory than the server. Would be nice to come up with a nicer solution but I'm "takki auki" here so to speak :(

File: test/e2e/specs/authentication-spec.js (hunk 3 above)
Comment (emilvirkki) [reply test/e2e/specs/authentication-spec.js:4]: Done :+1:

File: test/e2e/specs/participants-spec.js (hunk 4 above)
Comment (emilvirkki) [reply test/e2e/specs/participants-spec.js:14]: Makes sense, I'll add that.

File: test/utils/test-utils.js (hunk 5 above)
Comment (emilvirkki) [reply test/utils/test-utils.js:88]: Trying to delete some models and their dependencies at the same time causes intermittent database deadlocks at least on my machine, so you need to delete each type of model consecutively.

File: test/utils/test-utils.js (hunk 6 above)
Comment (emilvirkki) [reply test/utils/test-utils.js:74]: We need to create the fixtures in a particular order so that e.g. Participants are created before ParticipantDates, so the Promise.all() doesn't work. There could be more elegant approaches to the fixture creation though 🤔

Comment (googol) [reply test/utils/test-utils.js:74]: Oh right. This is a self-contained function so this for-loop isn't that bad really anyways. Using `Object.entries` would make the indexing a bit clleaner though:

```js
for (const [model, fixture] of Object.entries(fixtures)) {
  await createFixture(model, fixture)
}
```

**Comment by googol** [comment]
There's a conflict in package-lock.json where the master side has updated the package `bl` and your side has replaced it with `blob-util`. I'm guessing the `bl` package is no longer needed?

File: package.json [hunk 7]
```diff
@@ -25,16 +25,17 @@
     "test-watch": "cross-env TS_NODE_PROJECT=src/server/tsconfig.json NODE_ENV=test ENABLE_OFFLINE_LOGIN=true mocha --watch --timeout 4000 --require ts-node/register --exit",
     "test-integration": "cross-env TS_NODE_PROJECT=src/server/tsconfig.json NODE_ENV=test mocha --timeout 4000 --require ts-node/register test/integration --exit",
     "pretest-e2e": "cross-env NODE_ENV=test npm run seed-database && npm run build",
-    "test-e2e-mocha": "mocha --timeout 10000 --require ts-node/register test/e2e --exit",
-    "test-e2e": "cross-env TS_NODE_PROJECT=src/server/tsconfig.json NODE_ENV=test PORT=3005 start-test 'mkdir -p logs && npm start > ./logs/server-test.log 2>&1' 3005 test-e2e-mocha",
+    "test-e2e-cypress": "cross-env TS_NODE_PROJECT= cypress run --project ./test/e2e --browser chrome --headless",
```
[the remark this answers is not in the document]
Comment (googol) [reply package.json]: removed it
